CMakeLists.txt

```diff
@@ -1,85 +1,106 @@
 cmake_minimum_required(VERSION 3.3)
 
-project(rabit VERSION 0.3.0)
-
-include(CheckCXXCompilerFlag)
-CHECK_CXX_COMPILER_FLAG("-std=c++11" COMPILER_SUPPORTS_CXX11)
-if(COMPILER_SUPPORTS_CXX11)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
-else()
-  message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++11 support. Please use a different C++ compiler.")
-endif()
+project(rabit VERSION 0.3.0 LANGUAGES CXX)
 
 option(RABIT_BUILD_TESTS "Build rabit tests" OFF)
 option(RABIT_BUILD_MPI "Build MPI" OFF)
 option(RABIT_BUILD_DMLC "Include DMLC_CORE in build" OFF)
+option(RABIT_WITH_R_LIB "Fit the strict environment of R" OFF)
+
+option(DMLC_ROOT "Specify root of external dmlc core.")
+# by default point to xgboost/dmlc-core
+set(DMLC_ROOT ${CMAKE_CURRENT_LIST_DIR}/../dmlc-core)
```

Comment:
You spent so much effort on rabit I assume you don't want to make it XGBoost exclusive right?  Since last time I check rabit is only using some headers from `dmlc-core`, there's no linkage involved.  So technically rabit can use dmlc-core in system path, say `/usr/` or `/home/chenqin/.local/` or any other directories that follow the `bin/include/lib/lib64` pattern without any problem, and you can set above path as `DMLC_ROOT` for you convenience.  With this, you can simply throw an error if DMLC_ROOT is not specified.

In the future (not for this PR), I may add this to rabit as `dmlc-core` has very good support for CMake target.
``` cmake
if (NOT DMLC_ROOT)
  find_package(dmlc)
  target_link_libraries(rabit dmlc::dmlc)  # This line also handles include directory.
endif (NOT DMLC_ROOT)
```